wpiutil/src/main/native/include/wpi/jni_util.h

```diff
@@ -175,197 +175,276 @@ class JStringRef {
   SmallString<128> m_str;
 };
 
-// Details for J*ArrayRef and CriticalJ*ArrayRef
 namespace detail {
 
-template <typename C, typename T>
-class JArrayRefInner {
- public:
-  operator std::span<const T>() const {  // NOLINT
-    return static_cast<const C*>(this)->array();
-  }
+template <typename T>
+struct ArrayHelper {
+  static constexpr bool is_defined = false;
 };
 
+#define WPI_JNI_ARRAYHELPER(T, F)                                             \
+  template <>                                                                 \
+  struct ArrayHelper<T> {                                                     \
+    static constexpr bool is_defined = true;                                  \
+    using jarray_type = T##Array;                                             \
+    static T* GetArrayElements(JNIEnv* env, jarray_type jarr) {               \
+      return env->Get##F##ArrayElements(jarr, nullptr);                       \
+    }                                                                         \
+    static void ReleaseArrayElements(JNIEnv* env, jarray_type jarr, T* elems, \
+                                     jint mode) {                             \
+      env->Release##F##ArrayElements(jarr, elems, mode);                      \
+    }                                                                         \
+  };
+
+WPI_JNI_ARRAYHELPER(jboolean, Boolean)
+WPI_JNI_ARRAYHELPER(jbyte, Byte)
+WPI_JNI_ARRAYHELPER(jshort, Short)
+WPI_JNI_ARRAYHELPER(jint, Int)
+WPI_JNI_ARRAYHELPER(jlong, Long)
+WPI_JNI_ARRAYHELPER(jfloat, Float)
+WPI_JNI_ARRAYHELPER(jdouble, Double)
+
+#undef WPI_JNI_ARRAYHELPER
+
 /**
- * Specialization of JArrayRefBase to provide std::string_view conversion
- * and span<const uint8_t> conversion.
+ * Helper class for working with JNI arrays.
+ *
+ * This class exposes an is_valid() member and an explicit conversion to bool
+ * which indicate if the span is valid. Operations on invalid spans are
+ * undefined.
+ *
+ * Note that Set<PrimitiveType>ArrayRegion may be faster for pure writes since
+ * it avoids copying the elements from Java to C++.
+ *
+ * @tparam T The element type of the array (e.g., jdouble).
+ * @tparam IsCritical If true, Get/ReleasePrimitiveArrayCritical will be used
+ * instead of Get/Release\<PrimitiveType\>ArrayElements.
+ * @tparam Size The number of elements in the span.
  */
-template <typename C>
-class JArrayRefInner<C, jbyte> {
+template <typename T, bool IsCritical, size_t Size = std::dynamic_extent>
+  requires ArrayHelper<std::remove_cv_t<T>>::is_defined
```

Comment:
Instead of is_defined, can't you just look for the inner type jarray_type?

Reply:
Done, though it seems that C++ doesn't provide a built-in way to check if a type exists, so I had to add one.

Reply:
```c++
template<typename T>
concept IsJArray = requires
{
    typename T::jarray_type; // required nested member name
};
```